resonate/bridge.py

```diff
@@ -88,57 +89,55 @@ def __init__(
-        task: Task | None = None
-
-        if cmd.func is None:
-            # RPC case
-
-            promise = self._store.promises.create(
-                id=cmd.id,
-                ikey=cmd.id,
-                timeout=cmd.opts.timeout,
-                data={"func": cmd.name, "args": cmd.args, "kwargs": cmd.kwargs, "version": cmd.opts.version},
-                tags={
-                    **cmd.opts.tags,
-                    "resonate:invoke": cmd.opts.send_to,
-                    "resonate:scope": "global",
-                },
-            )
-        else:
-            # Invocation case
-            promise, task = self._store.promises.create_with_task(
-                id=cmd.id,
-                ikey=cmd.id,
-                timeout=cmd.opts.timeout,
-                pid=self._pid,
-                ttl=self._ttl * 1000,
-                data={"func": cmd.name, "args": cmd.args, "kwargs": cmd.kwargs, "version": cmd.opts.version},
-                tags={
-                    **cmd.opts.tags,
-                    "resonate:invoke": cmd.opts.send_to,
-                    "resonate:scope": "global",
-                },
-            )
-
-        futures[0].set_result(promise)
+    def run(self, rconv: RConv, lconv: LConv, future: Future) -> DurablePromise:
+        promise, task = self._store.promises.create_with_task(
+            id=rconv.id,
+            ikey=rconv.id,
+            timeout=rconv.timeout,
+            data=rconv.data,
+            tags={**rconv.tags, "resonate:scope": "global"},
+            pid=self._pid,
+            ttl=self._ttl * 1000,
+        )
 
-        # TODO(avillega): What happens if promise is completed?
-        if task is not None:
+        if promise.completed:
+            assert not task
+            match promise.result:
+                case Ok(v):
+                    future.set_result(v)
+                case Ko(e):
+                    future.set_exception(e)
+        elif task is not None:
             self._promise_id_to_task[promise.id] = task
             self.start_heartbeat()
-            self._cq.put_nowait((cmd, futures))
+            self._cq.put_nowait((Invoke(lconv.id, lconv.func, lconv.args, lconv.kwargs, lconv.opts), future))
+        else:
+            self._cq.put_nowait((Listen(promise.id), future))
+
+        return promise
+
+    def rpc(self, rconv: RConv, future: Future) -> DurablePromise:
+        promise = self._store.promises.create(
+            id=rconv.id,
+            ikey=rconv.id,
+            timeout=rconv.timeout,
+            data=rconv.data,
+            tags={**rconv.tags, "resonate:scope": "global"},
+        )
+
+        if promise.completed:
+            match promise.result:
+                case Ok(v):
+                    future.set_result(v)
+                case Ko(e):
+                    future.set_exception(e)
         else:
-            self._cq.put_nowait((Listen(promise.id), futures))
+            self._cq.put_nowait((Listen(promise.id), future))
```

Comment:
I think we might want to lazily put a `Listen` on the queue when the user calls result on the handle. What do you think @avillega?